src/ngx_http_lua_resolve.c

```diff
@@ -0,0 +1,483 @@
+    unsigned                      ipv6:1;
+};
+
+
+static int ngx_http_lua_ngx_resolve(lua_State *L)
+{
+    int                          n, saved_top;
+    size_t                       len;
+    u_char                      *p;
+    ngx_int_t                    rc;
+    ngx_str_t                    host;
+    ngx_url_t                    url;
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_http_request_t          *r;
+    ngx_resolver_ctx_t          *rctx, temp;
+    ngx_http_lua_co_ctx_t       *coctx;
+    ngx_http_core_loc_conf_t    *clcf;
+    ngx_http_lua_resolve_ctx_t  *u;
+
+    unsigned  ipv4 = 1;
+    unsigned  ipv6 = 0;
+
+    n = lua_gettop(L);
+    if (n != 1 && n != 2) {
+        return luaL_error(L, "passed %d arguments, but accepted 1 or 2", n);
+    }
+
+    if (lua_type(L, 1) != LUA_TSTRING) {
+        return luaL_error(L, "1st parameter must be a string");
+    }
+
+    if (n == 2) {
+        if (lua_type(L, 2) != LUA_TTABLE) {
+            return luaL_error(L, "2nd parameter must be a table");
+        }
+
+        rc = ngx_http_lua_resolve_set_query_option("ipv4", &ipv4, L);
+        if (rc != NGX_OK) {
+            return rc;
+        }
+
+        rc = ngx_http_lua_resolve_set_query_option("ipv6", &ipv6, L);
+        if (rc != NGX_OK) {
+            return rc;
+        }
+    }
+
+    r = ngx_http_lua_get_req(L);
+    if (r == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return luaL_error(L, "no ctx found");
+    }
+
+    ngx_http_lua_check_context(L, ctx, NGX_HTTP_LUA_CONTEXT_REWRITE
+                                       | NGX_HTTP_LUA_CONTEXT_ACCESS
+                                       | NGX_HTTP_LUA_CONTEXT_CONTENT);
+
+    p = (u_char *) luaL_checklstring(L, 1, &len);
+
+    host.data = ngx_palloc(r->pool, len + 1);
+    if (host.data == NULL) {
+        return luaL_error(L, "no memory");
+    }
+
+    host.len = len;
+
+    ngx_memcpy(host.data, p, len);
+    host.data[len] = '\0';
+
+    ngx_memzero(&url, sizeof(ngx_url_t));
+
+    url.url.len = host.len;
+    url.url.data = host.data;
+    url.default_port = (in_port_t) 0;
+    url.no_resolve = 1;
+
+    if (ngx_parse_url(r->pool, &url) != NGX_OK) {
+        lua_pushnil(L);
+
+        if (url.err) {
+            lua_pushfstring(L, "failed to parse host name \"%s\": %s",
+                            host.data, url.err);
+
+        } else {
+            lua_pushfstring(L, "failed to parse host name \"%s\"", host.data);
+        }
+
+        return 2;
+    }
+
+    if (url.addrs && url.addrs[0].sockaddr) {
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+            "the given host name is address");
+
+        return ngx_http_lua_resolve_get_retval(url.addrs[0].sockaddr,
+                                               url.addrs[0].socklen, L);
+    }
+
+    u = ngx_pcalloc(r->pool, sizeof(ngx_http_lua_resolve_ctx_t));
```

Comment:
It's not a good idea to allocate in the request pool. If the user calls this `resolve()` Lua function in a tight Lua loop, for example, then there will be a horrible memory leak. Providing an OO-like interface similar to `ngx.pipe` and cosockets might make more sense.

Reply:
Thanks for that point. I completely reorganized memory management approach in my new implementation.